urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+                                   (longtext.length() + 3) / 4);
+    // Check again based on the third quarter.
+    String[] hm2 = diff_halfMatchI(longtext, shorttext,
+                                   (longtext.length() + 1) / 2);
+    String[] hm;
+    if (hm1 == null && hm2 == null) {
+      return null;
+    } else if (hm2 == null) {
+      hm = hm1;
+    } else if (hm1 == null) {
+      hm = hm2;
+    } else {
+      // Both matched.  Select the longest.
+      hm = hm1[4].length() > hm2[4].length() ? hm1 : hm2;
+    }
+
+    // A half-match was found, sort out the return data.
+    if (text1.length() > text2.length()) {
+      return hm;
+      //return new String[]{hm[0], hm[1], hm[2], hm[3], hm[4]};
+    } else {
+      return new String[]{hm[2], hm[3], hm[0], hm[1], hm[4]};
+    }
+  }
+
+  /**
+   * Does a substring of shorttext exist within longtext such that the
+   * substring is at least half the length of longtext?
+   * @param longtext Longer string.
+   * @param shorttext Shorter string.
+   * @param i Start index of quarter length substring within longtext.
+   * @return Five element String array, containing the prefix of longtext, the
+   *     suffix of longtext, the prefix of shorttext, the suffix of shorttext
+   *     and the common middle.  Or null if there was no match.
+   */
+  private String[] diff_halfMatchI(String longtext, String shorttext, int i) {
+    // Start with a 1/4 length substring at position i as a seed.
+    String seed = longtext.substring(i, i + longtext.length() / 4);
+    int j = -1;
+    String best_common = "";
+    String best_longtext_a = "", best_longtext_b = "";
+    String best_shorttext_a = "", best_shorttext_b = "";
+    while ((j = shorttext.indexOf(seed, j + 1)) != -1) {
+      int prefixLength = diff_commonPrefix(longtext.substring(i),
+                                           shorttext.substring(j));
+      int suffixLength = diff_commonSuffix(longtext.substring(0, i),
+                                           shorttext.substring(0, j));
+      if (best_common.length() < suffixLength + prefixLength) {
+        best_common = shorttext.substring(j - suffixLength, j)
+            + shorttext.substring(j, j + prefixLength);
+        best_longtext_a = longtext.substring(0, i - suffixLength);
+        best_longtext_b = longtext.substring(i + prefixLength);
+        best_shorttext_a = shorttext.substring(0, j - suffixLength);
+        best_shorttext_b = shorttext.substring(j + prefixLength);
+      }
+    }
+    if (best_common.length() * 2 >= longtext.length()) {
+      return new String[]{best_longtext_a, best_longtext_b,
+                          best_shorttext_a, best_shorttext_b, best_common};
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Reduce the number of edits by eliminating semantically trivial equalities.
+   * @param diffs LinkedList of Diff objects.
+   */
+  public void diff_cleanupSemantic(LinkedList<Diff> diffs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 908-958 into a separate method.